web-api/src/storage/elastic/client.rs

```diff
@@ -0,0 +1,315 @@
+    ) -> Result<Option<T>, Error>
+    where
+        B: Into<Body>,
+        T: DeserializeOwned,
+    {
+        let request_builder = if let Some((headers, body)) = post_data {
+            self.client.post(url).headers(headers).body(body)
+        } else {
+            self.client.get(url)
+        };
+
+        let response = request_builder
+            .basic_auth(
+                &self.config.user,
+                Some(self.config.password.expose_secret()),
+            )
+            .send()
+            .await?;
+
+        let status = response.status();
+        if status == StatusCode::NOT_FOUND {
+            Ok(None)
+        } else if !status.is_success() {
+            let url = response.url().clone();
+            let body = response.bytes().await?;
+            let error = String::from_utf8_lossy(&body).into_owned();
+            Err(Error::Status { status, url, error })
+        } else {
+            Ok(Some(response.json().await?))
+        }
+    }
+
+    pub(super) async fn query_with_json<B, T>(
+        &self,
+        url: Url,
+        body: Option<B>,
+    ) -> Result<Option<T>, Error>
+    where
+        B: Serialize,
+        T: DeserializeOwned,
+    {
+        let post_data = body
+            .map(|json| -> Result<_, Error> {
+                let mut headers = HeaderMap::new();
+                headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
+                let body = serde_json::to_vec(&json)?;
+                Ok((headers, body))
+            })
+            .transpose()?;
+
+        self.query_with_bytes(url, post_data).await
+    }
+}
+
+#[derive(Debug, Error, displaydoc::Display, From)]
+pub(crate) enum Error {
+    /// Transmitting a request or receiving the response failed: {0}
+    Transport(reqwest::Error),
+    /// Elastic Search failed, status={status}, url={url}, \nbody={error}
+    Status {
+        status: StatusCode,
+        url: Url,
+        error: String,
+    },
+    /// Failed to serialize a requests or deserialize a response.
+    Serialization(serde_json::Error),
+    /// Given endpoint was not found: {0}
+    EndpointNotFound(&'static str),
+}
+
+#[derive(derive_more::Into, Clone)]
+struct SegmentableUrl(Url);
+
+impl SegmentableUrl {
+    fn with_segments(&self, segments: impl IntoIterator<Item = impl AsRef<str>>) -> Self {
+        let mut new_url = self.0.clone();
+        let mut segments_mut = new_url.path_segments_mut()
+            .unwrap(/* we made sure this can't happen */);
+        for segment in segments {
+            segments_mut.push(segment.as_ref());
+        }
+        drop(segments_mut);
+        Self(new_url)
+    }
+}
+
+impl TryFrom<Url> for SegmentableUrl {
+    type Error = anyhow::Error;
+
+    fn try_from(mut url: Url) -> Result<Self, Self::Error> {
+        url.path_segments_mut()
+            .map_err(|()| anyhow::anyhow!("non segmentable url"))?;
+        Ok(Self(url))
+    }
```

Comment:
🔧 since we don't actually need the segments here, but instead just want to test for it, we could do the test right away:
```suggestion
    fn try_from(url: Url) -> Result<Self, Self::Error> {
        if url.cannot_be_a_base() {
            Err(anyhow::anyhow!("non segmentable url"))
        } else {
            Ok(Self(url))
        }
    }
```